File: ios/Utils/ClusterProcessor.swift
//
//  ClusterProcessor.swift
//  Glimpse
//
//  Created by Nitin Issac Joy on 10/07/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation
import RealmSwift

class ClusterProcessor {
  
  static func RunStepClustering(trip: [ClusterModal]) -> [StepModal] {
    if(trip.count == 0) {
       return [StepModal()]
    }
    
    var stepResult: [StepModal] = []
    var firstTimestamp: Int64 = trip[0].timestamp
    var firstItem: ClusterModal = trip[0]
    
    var _stepCluster: [[ClusterModal]] = []
    _stepCluster.append([])
    
    var _it: Int = 0
    
    for item in trip {
      // If distance between two modals are less than 10 km
      if(ClusterProcessor.EarthDistance(p: item, q: firstItem) < 10) {
        // We divide based on time
        if(item.timestamp <= firstTimestamp + 86400) {
          _stepCluster[_it].append(item)
        } else {
          firstTimestamp = item.timestamp
          _stepCluster.append([item])
          _it += 1;
        }
      } else {
        // Else we divide based on distance
        firstTimestamp = item.timestamp
        firstItem = item;
        _stepCluster.append([item])
        _it += 1;
      }
    }
    
    for cluster in _stepCluster {
      let _step = ClusterProcessor.convertClusterToStep(cluster: cluster)
      if(_step.stepId != -1) {
        stepResult.append(_step)
      }
    }
    
    stepResult = stepResult.sorted(by: { $0.endTimestamp < $1.endTimestamp } )
    
    var i = 100;
    var previousStep: StepModal = StepModal();
    var distanceTravelled = 0
    for step in stepResult {
      step.stepId = i
      if(i > 100) {
        
        let _m = ClusterModal()
        _m.latitude = step.meanLatitude
        _m.longitude = step.meanLongitude
        
        let _n = ClusterModal()
        _n.latitude = previousStep.meanLatitude
        _n.longitude = previousStep.meanLongitude
        distanceTravelled += ClusterProcessor.EarthDistance(p: _m , q: _n)
      }
      step.distanceTravelled = distanceTravelled;
      step.desc = "Description goes here...";
      i += 100;
      previousStep = step;
    }
    
    return stepResult;
  }
  
  static func RunMasterClustering(clusterData: [ClusterModal], homes: List<HomesForDataClusteringModal>) -> [[ClusterModal]] {
  
    var trips: [[ClusterModal]] = []
    var trip: [ClusterModal] = []
  
    var _clusterData = clusterData.sorted(by: { $0.timestamp < $1.timestamp })
  
    var prevData: ClusterModal = _clusterData[0];
    for data in _clusterData {
      let _clusterHomeData = ClusterModal()
      _clusterHomeData.latitude = homes[Int(data.timestamp/86400)].latitude
      _clusterHomeData.longitude = homes[Int(data.timestamp/86400)].longitude
      _clusterHomeData.timestamp = homes[Int(data.timestamp/86400)].timestamp
      
      // If distance from home is more than 40 km
      if(ClusterProcessor.EarthDistance(p: _clusterHomeData, q: data) > 40
        // Noise filtering, if two pictures are taken 7 days apart, consider it a new trip
        && (ClusterProcessor.TimeDistance(p: data, q: prevData) < 86400*7)) {
          trip.append(data)
      } else if(trip.count > 0){
          // If more than 3 photo were taken for the trip, it's officially considered a trip
          if(trip.count > 3) {
            trips.append(trip)
          }
          trip = []
      }
      prevData = data
    }
  
    if(trip.count > 0) {
      trips.append(trip)
    }
    return trips;
  }
  
  static func TimeDistance(p: ClusterModal, q: ClusterModal) -> Int64 {
    return abs(p.timestamp - q.timestamp)
  }
  
  static func EarthDistance(p: ClusterModal, q: ClusterModal) -> Int {
    
    let lat2 = q.latitude;
    let lat1 = p.latitude;
    let lon2 = q.longitude;
    let lon1 = p.longitude;
  
    let R: Float64 = 6371; // km
    let dLat = ClusterProcessor.deg2rad(deg: lat2 - lat1);
    let dLon = ClusterProcessor.deg2rad(deg: lon2 - lon1);
    let a = sin(dLat / 2) * sin(dLat / 2) +
      cos(ClusterProcessor.deg2rad(deg: lat1)) * cos(ClusterProcessor.deg2rad(deg: lat2)) * sin(dLon / 2) * sin(dLon / 2);
    let c = 2 * atan2(a.squareRoot(), (1 - a).squareRoot());
    let d = R * Float64(c);
    return Int(d);
  }
  
  static func convertClusterToStep(cluster: [ClusterModal]) -> StepModal {
    if(cluster.count == 0) {
      let _step = StepModal()
      _step.stepId = -1
      return _step
    }
    
    var latitudeSum: Float64 = 0;
    var longitudeSum: Float64 = 0;
    var imageUris: [String] = []
    var markers: [Region] = []
    
    var _cluster = cluster.sorted(by: { return $0.timestamp < $1.timestamp })
    
    for item in _cluster {
      latitudeSum += item.latitude;
      longitudeSum += item.longitude;
      imageUris.append(item.image);
      let _r = Region()
      _r.latitude = item.latitude;
      _r.longitude = item.longitude;
      markers.append(_r)
    }
    
    let _step: StepModal = StepModal()
    _step.meanLatitude = latitudeSum/Double(cluster.count);
    _step.meanLongitude = longitudeSum/Double(cluster.count);
    _step.markers = List<Region>()
    for marker in markers {
      _step.markers.append(marker)
    }
    _step.startTimestamp = _cluster[0].timestamp;
    _step.endTimestamp = _cluster[cluster.count - 1].timestamp
    _step.images = List<Image>()
    for image in imageUris {
      let _obj = Image()
      _obj.image = image
      _step.images.append(_obj)
    }
    _step.masterImage = imageUris[0];
    let _r = Region()
    _r.latitude = _step.meanLatitude
    _r.longitude = _step.meanLongitude
    _step.masterMarker = _r
    
    return _step;
  }

  static func deg2rad(deg: Float64) -> Float64 {
    return deg * (Double.pi/180)
  }
}

File: android/app/src/main/java/com/beerwithai/glimpse/Engine/Utils/ClusterProcessor.java
package com.beerwithai.glimpse.Engine.Utils;

import com.beerwithai.glimpse.Engine.Modals.ClusterModal;
import com.beerwithai.glimpse.Engine.Modals.HomesForDataClusteringModal;
import com.beerwithai.glimpse.Engine.Modals.Region;
import com.beerwithai.glimpse.Engine.Modals.StepModal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import io.realm.RealmList;

class CompareByTimestamp implements Comparator<StepModal> {
    public int compare(StepModal a, StepModal b) {
        return (int)(a.endTimestamp - b.endTimestamp);
    }
}

class CompareClusterByTimestamp implements  Comparator<ClusterModal> {
    public int compare(ClusterModal a, ClusterModal b) {
        return (int)(a.timestamp - b.timestamp);
    }
}

public class ClusterProcessor {


    public static ArrayList<StepModal> RunStepClustering(ArrayList<ClusterModal> trip) {
        if(trip.size() == 0) {
            return null;
        }

        ArrayList<StepModal> stepResult = new ArrayList<StepModal>();
        Double firstTimestamp = trip.get(0).timestamp;
        ClusterModal firstItem = trip.get(0);
        ArrayList<ArrayList<ClusterModal>> _stepCluster = new ArrayList<ArrayList<ClusterModal>>();
        _stepCluster.add(new ArrayList<ClusterModal>());
        Integer _it = 0;

        for(int i = 0; i < trip.size(); i++) {
            if(ClusterProcessor.EarthDistance(trip.get(i), firstItem) < 10) {
                if(trip.get(i).timestamp <= firstTimestamp + 86400) {
                    _stepCluster.get(_it).add(trip.get(i));
                } else {
                    firstTimestamp = trip.get(i).timestamp;
                    ArrayList<ClusterModal> t = new ArrayList<ClusterModal>();
                    t.add(trip.get(i));
                    _stepCluster.add(t);
                    _it += 1;
                }
            } else {
                firstTimestamp = trip.get(i).timestamp;
                firstItem = trip.get(i);
                ArrayList<ClusterModal> t = new ArrayList<ClusterModal>();
                t.add(trip.get(i));
                _stepCluster.add(t);
                _it += 1;
            }
        }

        for(int i = 0; i < _stepCluster.size(); i++) {
            StepModal step = ClusterProcessor.convertClusterToStep(_stepCluster.get(i));
            if(step.stepId != -1) {
                stepResult.add(step);
            }
        }

        Collections.sort(stepResult, new CompareByTimestamp());

        StepModal previousStep = new StepModal();
        int distanceTravelled = 0;
        for(int i = 0; i < stepResult.size(); i++) {
            stepResult.get(i).stepId = i*100;
            if(i > 1) {
                ClusterModal _m = new ClusterModal();
                _m.latitude = stepResult.get(i).meanLatitude;
                _m.longitude = stepResult.get(i).meanLongitude;

                ClusterModal _n = new ClusterModal();
                _n.latitude = previousStep.meanLatitude;
                _n.longitude = previousStep.meanLongitude;

                distanceTravelled += ClusterProcessor.EarthDistance(_m, _n);
            }
            stepResult.get(i).distanceTravelled = distanceTravelled;
            stepResult.get(i).desc = "Description goes here...";
            previousStep = stepResult.get(i);
        }

        return stepResult;
    }

    public static ArrayList<ArrayList<ClusterModal>> RunMasterClustering(ArrayList<ClusterModal> clusterData, ArrayList<HomesForDataClusteringModal> homes) {
        ArrayList<ArrayList<ClusterModal>> trips = new ArrayList<ArrayList<ClusterModal>>();
        ArrayList<ClusterModal> trip = new ArrayList<ClusterModal>();

        Collections.sort(clusterData, new CompareClusterByTimestamp());
        ClusterModal prevData = clusterData.get(0);

        for(int i = 0; i < clusterData.size(); i++) {
            ClusterModal _clusterHomeData  = new ClusterModal();
            _clusterHomeData.latitude = homes.get((int)(clusterData.get(i).timestamp/86400)).latitude;
            _clusterHomeData.longitude = homes.get((int)(clusterData.get(i).timestamp/86400)).longitude;
            _clusterHomeData.timestamp = homes.get((int)(clusterData.get(i).timestamp/86400)).timestamp;

            if(ClusterProcessor.EarthDistance(_clusterHomeData, clusterData.get(i)) > 40 && (ClusterProcessor.TimeDistance(clusterData.get(i), prevData) < 86400 * 7)) {
                trip.add(clusterData.get(i));
            } else if(clusterData.size() > 0) {
                if(clusterData.size() > 3) {
                    trips.add(trip);
                }
                trip = new ArrayList<ClusterModal>();
            }
            prevData = clusterData.get(i);
        }

        if(trip.size() > 0) {
            trips.add(trip);
        }

        return trips;
    }

    public static Double TimeDistance(ClusterModal p, ClusterModal q) {
        return Math.abs(p.timestamp - q.timestamp);
    }

    public static Integer EarthDistance(ClusterModal p, ClusterModal q) {
        double lat1 = p.latitude;
        double lat2 = q.latitude;
        double lon1 = q.longitude;
        double lon2 = q.longitude;

        double R = 6371;
        double dLat = ClusterProcessor.deg2rad(lat2 - lat1);
        double dLon = ClusterProcessor.deg2rad(lon2 - lon1);

        double a = Math.sin(dLat/2) * Math.sin(dLon/2) + Math.cos(ClusterProcessor.deg2rad(lat1)) * Math.cos(ClusterProcessor.deg2rad(lat2)) * Math.sin(dLon/2) * Math.sin(dLon/2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));

        double d = R * Float.valueOf((float)c);

        return (int)d;
    }

    public static double deg2rad(double deg) {
        return deg * (Math.PI/180);
    }

    public static StepModal convertClusterToStep(ArrayList<ClusterModal> cluster) {
        if(cluster.size() == 0) {
            StepModal step = new StepModal();
            step.stepId = -1;
            return step;
        }

        double latitudeSum = 0;
        double longitudeSum = 0;
        ArrayList<String> imageUris = new ArrayList<String>();
        ArrayList<Region> markers = new ArrayList<Region>();

        Collections.sort(cluster, new CompareClusterByTimestamp());
        for(int i = 0; i < cluster.size(); i++) {
            latitudeSum += cluster.get(i).latitude;
            longitudeSum += cluster.get(i).longitude;
            imageUris.add(cluster.get(i).image);

            Region _r = new Region();
            _r.latitude = cluster.get(i).latitude;
            _r.longitude = cluster.get(i).longitude;

            markers.add(_r);
        }

        StepModal step = new StepModal();
        step.meanLatitude = Float.valueOf((float)(latitudeSum/cluster.size()));
        step.meanLongitude = Float.valueOf((float)longitudeSum/cluster.size());
        step.markers = new RealmList<Region>();

        for(int i = 0; i < markers.size(); i++) {
            step.markers.add(markers.get(i));
        }

        step.startTimestamp = cluster.get(0).timestamp;
        step.endTimestamp = cluster.get(cluster.size() - 1).timestamp;
        step.images = new RealmList<String>();

        for(int i = 0; i < imageUris.size(); i++) {
            step.images.add(imageUris.get(i));
        }

        step.masterImage = imageUris.get(0);

        return step;
    }

}
